dracclient/tests/test_raid.py

```diff
@@ -112,6 +113,216 @@ def setUp(self):
+    def test_get_raid_controller_mode(self, mock_requests,
+                                      mock_wait_until_idrac_is_ready):
+
+        mock_requests.post('https://1.2.3.4:443/wsman', [
+            {'text': test_utils.RAIDEnumerations[
+                uris.DCIM_RAIDEnumeration]['ok']},
+            {'text': test_utils.RAIDEnumerations[
+                uris.DCIM_RAIDString]['ok']}])
+        raid_cntrl_mode = self.drac_client.get_raid_controller_mode()
+        self.assertEqual('RAID', raid_cntrl_mode)
+
+    @mock.patch.object(dracclient.client.WSManClient,
+                       'wait_until_idrac_is_ready', spec_set=True,
+                       autospec=True)
+    @mock.patch.object(dracclient.client.WSManClient,
+                       'invoke', spec_set=True,
+                       autospec=True)
+    def test_set_raid_controller_settings(self, mock_requests,
+                                          mock_invoke,
+                                          mock_wait_until_idrac_is_ready):
+
+        mock_requests.post('https://1.2.3.4:443/wsman', [
+            {'text': test_utils.RAIDEnumerations[
+                uris.DCIM_RAIDEnumeration]['ok']},
+            {'text': test_utils.RAIDEnumerations[
+                uris.DCIM_RAIDString]['ok']}])
+
+        mock_invoke.return_value = lxml.etree.fromstring(
+            test_utils.RAIDInvocations[uris.DCIM_RAIDService][
+                'SetAttributes']['ok'])
+
+        result = self.drac_client.set_raid_controller_settings(
+            {'RAIDRequestedControllerMode': 'RAID'},
+            self.raid_controller_fqdd)
+
+        self.assertEqual({'is_commit_required': True,
+                          'is_reboot_required': constants.RebootRequired.true
+                          },
+                         result)
+
+    @mock.patch.object(dracclient.client.WSManClient,
+                       'wait_until_idrac_is_ready', spec_set=True,
+                       autospec=True)
+    def test_set_raid_controller_settings_with_unknown_attr(
+            self, mock_requests, mock_wait_until_idrac_is_ready):
+        mock_requests.post('https://1.2.3.4:443/wsman', [
+            {'text': test_utils.RAIDEnumerations[
+                uris.DCIM_RAIDEnumeration]['ok']},
+            {'text': test_utils.RAIDEnumerations[
+                uris.DCIM_RAIDString]['ok']},
+            {'text': test_utils.RAIDInvocations[
+                uris.DCIM_RAIDService]['SetAttributes']['error']}])
+
+        self.assertRaises(exceptions.InvalidParameterValue,
+                          self.drac_client.set_raid_controller_settings,
+                          {'foo': 'bar'}, self.raid_controller_fqdd)
+
+    @mock.patch.object(dracclient.client.WSManClient,
+                       'wait_until_idrac_is_ready', spec_set=True,
+                       autospec=True)
+    def test_set_raid_controller_settings_with_unchanged_attr(
+            self, mock_requests, mock_wait_until_idrac_is_ready):
```

Comment:
This test case is really confusing.  Looking in the sample xml file, there is a RAIDdefaultWritePolicy for every virtual disk.  This makes me think that setting attributes by attribute name must be invalid, because how would the idrac know which attribute to change?  Does it change them all?  Does it error out?

Reply:
To avoid this confusion, it might be better to pick an attribute that occurs only once, like RAIDControllerBootMode or something like that.